vowpalwabbit/parse_example.cc

```diff
@@ -50,29 +50,113 @@ int read_features_string(vw* all, v_array<example*>& examples)
+      , _p_spelling_feature_group(nullptr)
+      , _p_dictionary_feature_group(nullptr)
+      , _current_example(current_example)
+      , _redefine_char_flag(redefine_char_flag)
+      , _redefine_char_map(redefine_char_map)
+      , _namespace_dictionaries_map(namespace_dictionaries_map)
+      , _affix_features(affix_features)
+      , _spelling_features(spelling_features)
+      , _hash_function(hash_fn)
+  {
+  }
+  inline unsigned char feature_group_char() const { return _feature_group_char; }
+  inline uint64_t feature_group_hash() const { return _hash; }
+  inline features& get_current_fg() { return get_feature_group(_p_current_feature_group, feature_group_char()); }
+  inline features& get_affix_fg() { return get_feature_group(_p_affix_feature_group, affix_namespace); }
+  inline features& get_spelling_fg() { return get_feature_group(_p_spelling_feature_group, spelling_namespace); }
+  inline features& get_dictionary_fg() { return get_feature_group(_p_dictionary_feature_group, dictionary_namespace); }
+  inline void set_current_fg(const VW::string_view& fg_name)
+  {
+    _feature_group_name = fg_name;
+    auto namespace_char = static_cast<unsigned char>(_feature_group_name[0]);
+    if (_redefine_char_flag) namespace_char = (*_redefine_char_map)[namespace_char];  // redefine feature_group_char
+    _feature_group_char = namespace_char;
+    _hash = _hash_function(_feature_group_name.begin(), _feature_group_name.length(), _hash_seed);
+    _p_current_feature_group = nullptr;  // reset cached feature group
+    _affix = (*_affix_features)[_feature_group_char];
+    _spelling = (*_spelling_features)[_feature_group_char];
+    _namespace_dictionaries = &(*_namespace_dictionaries_map)[feature_group_char()];
+  }
+  inline void set_current_fg()
+  {
+    _feature_group_name = " ";
+    _feature_group_char = ' ';
+    _hash = this->_hash_seed == 0 ? 0 : uniform_hash("", 0, this->_hash_seed);
+    _p_current_feature_group = nullptr;  // reset cached feature group
+    _affix = (*_affix_features)[_feature_group_char];
+    _spelling = (*_spelling_features)[_feature_group_char];
+    _namespace_dictionaries = &(*_namespace_dictionaries_map)[feature_group_char()];
+  }
+  inline const VW::string_view& get_fg_name() const { return _feature_group_name; }
+  inline uint64_t get_affix_features() const { return _affix; }
+  inline const bool is_spelling_feature_group() const { return _spelling; }
+  inline const bool dictionary_exists() const { return !_namespace_dictionaries->empty(); }
+  inline const namespace_dictionaries* get_dictionary() const { return _namespace_dictionaries; }
+
+private:
+  inline features& get_feature_group(features*& p_features_ref, unsigned char namespace_char)
+  {
+    if (p_features_ref == nullptr)
+    {
+      p_features_ref = &_current_example->feature_space[namespace_char];
+
+      if (p_features_ref->empty()) _current_example->indices.emplace_back(namespace_char);
+    }
+    return {*p_features_ref};
```

Comment:
The brackets here are going to cause a temp copy of features to be returned, should just be the ref itself no?
```suggestion
    return *p_features_ref;
```

Reply:
Yep.

Reply:
Probably will get optimized out.  Still no reason to have it.

Reply:
As it is right now it is returning a reference to a temporary object which is undefined behavior.